flow/aggregates.py

```diff
@@ -0,0 +1,463 @@
+                            'the aggregate instance. Expected a callable, '
+                            f'got {type(func)}.')
+
+
+class _AggregatesStore:
+    """This class holds the information of all the aggregates associated with
+    a :py:class:`Aggregate`.
+
+    This is a callable class which, when called, generates all the aggregates.
+    When iterated through it's instance, all the aggregates are yielded.
+
+    :param aggregate:
+        Aggregate object associated with this class.
+    :type aggregate:
+        :py:class:`Aggregate`
+    :param jobs:
+        The signac job handles
+    :type jobs:
+        Iterator
+    """
+    def __init__(self, aggregate, jobs):
+        self._aggregate = aggregate
+
+        # We need to register the aggregates for this instance using the
+        # jobs provided.
+        self._aggregates = list()
+        self._aggregate_ids = dict()
+        self._register_aggregates(jobs)
+
+    def __iter__(self):
+        yield from self._aggregates
+
+    def __getitem__(self, id):
+        "Return an aggregate "
+        try:
+            return self._aggregate_ids[id]
+        except KeyError:
+            raise LookupError(f'Unable to find the aggregate having id {id} in '
+                              'the FlowProject')
+
+    def __contains__(self, aggregate):
+        """Return whether an aggregate is stored in the this
+        instance of :py:class:`_AggregateStore`
+
+        :param aggregate:
+            Aggregate of jobs
+        :type jobs:
+            tuple of :py:class:`signac.contrib.job.Job`
+        """
+        if aggregate in self._aggregates:
+            return True
+        else:
+            return False
+
+    def __len__(self):
+        return len(self._aggregates)
+
+    def __eq__(self, other):
+        return type(self) == type(other) and \
+               self._aggregates == other._aggregates and \
+               self._aggregate == other._aggregate
+
+    def __hash__(self):
+        blob = str(hash(self._aggregate))
+        return int(sha1(blob.encode('utf-8')).hexdigest(), 16)
+
+    def _register_aggregates(self, jobs):
+        """If the instance of this class is called then we will
+        generate aggregates and store them in self._aggregates
+        """
+        if isinstance(jobs, list):  # Got a tuple of jobs
+            project = jobs[0]._project
+        else:  # Got a project
+            project = jobs
+
+        aggregated_jobs = list(jobs)
+        if self._aggregate._select is not None:
+            aggregated_jobs = list(filter(self._aggregate._select, aggregated_jobs))
+        if self._aggregate._sort_by is not None:
+            aggregated_jobs = list(sorted(aggregated_jobs,
+                                          key=lambda job: job.sp[self._aggregate._sort_by],
+                                          reverse=bool(self._aggregate._reverse_order)))
+        aggregated_jobs = self._aggregate._aggregator([job for job in aggregated_jobs])
```

Comment:
Here you are converting to a list again. We don't need to convert to list multiple times. `sorted` always returns a list, and `filter` does not, but if we pass it into `sorted` it will return a list.